### DISCUSSION TITLE
FAQs by Jax

### DISCUSSION THREAD

**Discussion by jwoehr** [Qiskit/.github#22] [FAQ]
Qiskit Slack FAQs by [Jax](https://qiskit.slack.com/team/UBN94LFDF)

Much of this material may be **OUT OF DATE** nowadays.

**Comment by jwoehr** [comment]
Q. Is there a way I can use the qasm assembly file to generate the dag directly without converting it into a circuit and then  a dag from the circuit?
A. If you have loaded your qasm code into a Python str `my_qasm_str` then:
```
from qiskit.qasm.qasm import Qasm
from qiskit.converters import ast_to_dag
qasm = Qasm(data=my_qasm_str)
ast = qasm.parse()
dag = ast_to_dag(ast)
```

**Comment by jwoehr** [comment]
Q: My job failed with Internal Error [[9999]](https://quantum-computing.ibm.com/docs/admin/manage/errors#error9999). The manual says: "Contact IBM Quantum via [Slack](https://ibm.co/joinqiskitslack) for help.",
A: Error 9999 is the catchall “Something Went Wrong” error :slightly_smiling_face:
Usually seems to mean a service outage, planned or unplanned, or a server error.
First debug step: Wait and try again later.
And sometimes we ask [@francabrera](https://qiskit.slack.com/team/UDDANNYUE) or [@Jessie Yu](https://qiskit.slack.com/team/UBHA5HX2Q) or [@Paul Nation](https://qiskit.slack.com/team/U9W7DQJSE).
**If you ask** be sure to include the **Job Number** from the failed job in your posting!

**Comment by jwoehr** [comment]
Q: I have a sophisticated quantum computation / quantum mechanical question relative to Qiskit that is hard to get answered in Qiskit Slack. Where do I go?
A: Post your question on [Stack Exchange Quantum Computing](https://quantumcomputing.stackexchange.com/). Be sure to add the following tags: `ibm ibm-q-experience qiskit`. IBM engineers pay attention to messages tagged in this fashion.

**Comment by HuangJunye** [comment]
Thanks for making the FAQs. They are very useful!

**Comment by jwoehr** [comment]
Q: I am playing around with FakeArmonk and I was wondering if there was a way to specify a custom noise model. I am trying to modify it to have a [noise-free | custom noise] PulseSimulation, if that's something possible.

A:  First, see https://qiskit.org/documentation/tutorials/circuits_advanced/10_pulse_simulator_backend_model.html

There is also this comment by Abdullah Ash-Saki in [this Qiskit Slack posting](https://qiskit.slack.com/archives/CJG2E9LQP/p1616732142069600?thread_ts=1616727960.069400&cid=CJG2E9LQP)

> You can see all IBM Backends and can download their calibration data i.e., noise values as a csv file from your https://quantum-computing.ibm.com/ account.
> 
> Also, you can extract the noise values from Fake backends as well.
> ```
> from qiskit.test.mock import FakeMontreal
> backend = FakeMontreal()
> properties = backend.properties()
> two_q_gate_error_0_1 = properties.gate_error('cx', [0,1])
> ```
> The above snippet will give you the 2Q gate error of CX(0,1) for the FakeMontreal backend. Basically, properties variable will contain other noise values as well, and you can see in your jupyter notebook with a little bit of coding.

Q: But concerning the fact that you can read the error values like the gate_error or the readout error, can you set manually these values to a specific one?

A: Look at `qiskit-terra/qiskit/test/mock/backends/armonk/fake_armonk.py`
You write json files that give the properties you want.
Look at `qiskit-terra/qiskit/test/mock/backends/armonk/props_armonk.json`
E.g., `"name": "gate_error", "unit": "", "value": 0.00019769550670970334`

_( Also, see https://qiskit.org/documentation/stubs/qiskit.providers.aer.PulseSimulator.html#qiskit.providers.aer.PulseSimulator )_

**Comment by jwoehr** [comment]
Q: Can any one give me a good tutorial for learning Open Pulse?
A: Here are a few helpful links

- [Calibrating Qubits with Qiskit Pulse](https://qiskit.org/textbook/ch-quantum-hardware/calibrating-qubits-pulse.html)
- [Pulse gates](https://quantum-computing.ibm.com/lab/files/qiskit-tutorials/qiskit/circuits_advanced/05_pulse_gates.ipynb)
- [Qiskit Pulse: Programming Quantum Computers Through the Cloud with Pulses](https://www.youtube.com/watch?v=V_as5PufUiU) (Video)

**Comment by jwoehr** [comment]
Q: How do I perform boolean logic on a quantum circuit?
A: See:
- [BooleanExpression](https://qiskit.org/documentation/stubs/qiskit.circuit.classicalfunction.BooleanExpression.html#booleanexpression)
- [Lab 1. Quantum Circuits](https://qiskit.org/textbook/ch-labs/Lab01_QuantumCircuits.html)

**Comment by jwoehr** [comment]
Q: I need more access / more qubits on IBM Quantum. Whom do I ask?
A: See [Apply for IBM Quantum Credits](https://quantum-computing.ibm.com/credits-program).

**Comment by jwoehr** [comment]
Q: How do I start learning about Quantum Computing with Qiskit?
A: Here is a possible learning pathway:
1. Create an account on [IBM Quantum](https://quantum-computing.ibm.com/) and try out the drag-and-drop [Composer](https://quantum-computing.ibm.com/composer/).
2. [Qiskit Documentation Tutorials](https://qiskit.org/documentation/tutorials.html) is a good start.
3. [Learn Quantum Computation using Qiskit](https://qiskit.org/textbook/preface.html) is **_The_** Textbook.
4. [YouTube Qiskit channel](https://www.youtube.com/qiskit) offers video self-learning.

**Comment by jwoehr** [comment]
Q: I tried to [sign in to IBM Quantum](https://quantum-computing.ibm.com) using _{Google | some other account that I thought was supported}_ and it just returns me to the login page. What do I do?
A: Try [creating an IBM id](https://www.ibm.com/account/reg/signup) and using that to log in.
_... or ..._
Q: I was logged in {before | yesterday | a few days ago} to IBM Quantum but now when I try to go to IBM Quantum, my web browser can't connect to the site.
A: You've got an IBM id and your login went stale.

- Delete all IBM cookies and stored data from your browsers.
- Close and reopen your browser.
- Login to [MyIBM](https://myibm.ibm.com) using your IBM id.
- Then try IBM Quantum again

**Comment by jwoehr** [comment]
Q. How do I pursue an IBM Quantum Developer Certification?
A. [Learn More About the IBM Quantum Developer Certification — And How to Take it For Free](https://medium.com/qiskit/learn-more-about-the-ibm-quantum-developer-certification-and-how-to-take-it-for-free-d237f9765dc5)

**Comment by jwoehr** [comment]
Q. I found a mistake in the [Qiskit textbook _Learn Quantum Computation using Qiskit_](https://qiskit.org/textbook/preface.html)  . Where do I report the mistake?
A.[ Please file an Issue](https://github.com/qiskit-community/qiskit-textbook/issues). If the problem is with [the new version of the textbook](https://qiskit.org/textbook-beta/), [please file the issue here](https://github.com/Qiskit/platypus/issues).

**Comment by jwoehr** [comment]
Q: I installed Qiskit, and then I cloned the `qiskit-xxxxx` repository, but when I try to import, I get `No module named 'qiskit-xxxxx'`.
A: You need to install the package from the cloned repository. It's not enough that you cloned it, it must be individually installed.

1. Activate your Qiskit virtual environment
2. `cd` to the cloned directory
3. `pip install .`

Note: The "dot" `.` in `pip install .` means "the package rooted in this current directory".

**Reply by HuangJunye** [reply]
You can also install by `pip install git+[git URL]` or just do `pip install [qiskit-xxxx]`. Qiskit nature example:

```
pip install git+https://github.com/Qiskit/qiskit-nature.git
```
or
```
pip install qiskit-nature
```

**Reply by jwoehr** [reply]
@frankharkins on installing the classing Qiskit Textbook:

> To install the qiskit_textbook package, you can do
pip install git+https://github.com/qiskit-community/qiskit-textbook.git#subdirectory=qiskit-textbook-src 
(or another method in the [readme](http://pip%20install%20git+https//github.com/qiskit-community/qiskit-textbook.git#subdirectory=qiskit-textbook-src)). This package is separate from Qiskit and has caused a fair bit of confusion, so we're trying to move away from it and get things directly integrated into Qiskit where appropriate.
I wouldn't recommend building the old textbook locally as it's a bit archaic now, if you want to build the textbook, or contribute anything other than content changes then please use the new beta version at https://github.com/Qiskit/platypus. (edited)

**Comment by jwoehr** [comment]
Q: Is it true that the more I use a specific backend, the longer I have to wait due to my queue time by fair share?
A: No.  The scheduling is now global and is (mostly) backend independent.  Your wait time is determined by your usage amount relative to others waiting for that system.  "Your usage" here is your usage over all systems.

**Comment by jwoehr** [comment]
Q. How do I save a compiled quantum circuit to load it later?
A. Use [QPY serialization](https://qiskit.org/documentation/apidoc/qpy.html?highlight=qpy#module-qiskit.qpy).

**Comment by jwoehr** [comment]
Q. When I have a problem, should I post a screen shot or attach my commands and the output to my question in Slack or Stack Exchange?
A. Cut and paste or attach as a file your commands and their output and errors. Screen shots are often hard to read and miss critical lines of a large amount of output. Also, we can't copy your screenshot and paste the code from it into a file to test your code!

**Comment by jwoehr** [comment]
Q. What is the limit, the maximum number of circuits, for a batch job?
A. It may vary from access provider to access provider.  See the [System Configuration documentation](https://quantum-computing.ibm.com/lab/docs/iql/manage/systems/configuration) and the heading **Max circuits**

**Comment by jwoehr** [comment]
Q. My job received error code _XXXX_ (4-digit numeric code) ... what does this mean?
A. See [IBM Quantum Lab API error codes](https://quantum-computing.ibm.com/lab/docs/iql/manage/errors)

**Comment by jwoehr** [comment]
Q: I'm trying to use the Aer statevector simulator with my GPU but it's not recognized.

A. By default all simulation methods run on the CPU, however select methods also support running on a GPU if qiskit-aer was installed with GPU support on a compatible NVidia GPU and CUDA version. See https://qiskit.org/documentation/stubs/qiskit.providers.aer.AerSimulator.html#qiskit.providers.aer.AerSimulator

The aer-gpu simulator must be installed via `pip install qiskit-aer-gpu` See https://pypi.org/project/qiskit-aer-gpu/

**Comment by jwoehr** [comment]
Q: Does Qiskit have a way of simulating adaptive circuits? I.e. circuits with measurement and feedback?

A: See https://quantum-computing.ibm.com/services/resources/docs/resources/manage/systems/dynamic-circuits/introduction

**Comment by jwoehr** [comment]
Q: How do I dump a `QuantumCircuit` instance to OpenQASM3 in a file?

A: Assuming that you have a `QuantumCircuit` instance stored in the variable `my_circuit` which you wish to dump to `myfile.qasm` you can do something like this:
```
output_file_handle = open('myfile.qasm', "w")
strio = StringIO()
qasm3.dump(my_circuit, strio)
output_file_handle.write(strio.getvalue() + "\n")
output_file_handle.close()
```

**Comment by jwoehr** [comment]
Q: Where did Aqua and my favorite Aqua algorithm go?

A: Qiskit Aqua is no longer supported.  Please see the [Qiskit Aqua Migration Guide](https://github.com/qiskit-community/qiskit-aqua#migration-guide). Also, further migration may have taken place even after that: see, for example, [this Qiskit RFC](https://github.com/Qiskit/RFCs/pull/44#pullrequestreview-1589840804)
